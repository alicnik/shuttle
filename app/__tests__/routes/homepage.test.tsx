import { createMockFormData, setupRemixStub } from '../test-utils';
import Homepage, { action, links, loader, meta } from '../../routes/_index';
import { useActionData, useLoaderData, useNavigation } from '@remix-run/react';
import { fromPartial } from '@total-typescript/shoehorn';
import { act, screen } from '@testing-library/react';
import { createUser } from '~/models/user.server';
import { getInboxes, syncSession } from '~/lib/session.server';

vi.mock('~/lib/session.server');
vi.mock('~/models/user.server');
vi.mock('~/lib/generate-random-name.server', () => ({
  generateRandomName: () => 'mock-random-name',
}));
vi.mock('../../../db', () => ({ db: {} }));

describe('homepage', () => {
  beforeEach(async () => {
    const loaderData = await loader(fromPartial({}));
    vi.mocked(useLoaderData<typeof loader>).mockReturnValue(
      await loaderData.json()
    );
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it('renders the expected meta data', () => {
    const metaData = meta(fromPartial({}));

    expect(metaData).toEqual([
      { title: 'Shuttle' },
      {
        name: 'description',
        content: 'A temporary email service that minimises the back and forth.',
      },
    ]);
  });

  it('renders the expected links', () => {
    const linksData = links();

    expect(linksData).toEqual([
      {
        rel: 'preconnect',
        href: 'https://fonts.gstatic.com',
        crossOrigin: 'anonymous',
      },
      { rel: 'preconnect', href: 'https://fonts.googleapis.com' },
      {
        rel: 'stylesheet',
        href: 'https://fonts.googleapis.com/css2?family=Zen+Dots&display=swap',
      },
    ]);
  });

  it('should match the snapshot', () => {
    const { container } = setupRemixStub(<Homepage />);

    expect(container).toMatchSnapshot();
  });

  it("animates the shuttle logo when it's clicked", async () => {
    vi.useFakeTimers({ shouldAdvanceTime: true });

    const { user } = setupRemixStub(<Homepage />, {
      userOptions: {
        advanceTimers: vi.advanceTimersByTime,
      },
    });

    const image = screen.getByRole('img');

    await user.click(image);

    expect(image).toHaveClass('shuttle-logo-animation');

    act(() => vi.advanceTimersByTime(1250));

    expect(image).not.toHaveClass('shuttle-logo-animation');

    vi.runOnlyPendingTimers();
    vi.useRealTimers();
  });

  it('should render the input with an empty default value on first mount', () => {
    setupRemixStub(<Homepage />);

    const input = screen.getByRole('textbox');

    expect(input).toHaveValue('');
  });

  it('renders the input with a default value if a random name has been generated', async () => {
    const actionData = await action(
      fromPartial({
        request: {
          formData: async () => createMockFormData({ _action: 'random' }),
        },
      })
    );

    vi.mocked(useActionData<typeof action>).mockReturnValue(
      await actionData.json()
    );

    setupRemixStub(<Homepage />);

    expect(screen.getByRole('textbox')).toHaveValue('mock-random-name');
  });

  it('renders an error if the username is empty', async () => {
    vi.mocked(createUser).mockResolvedValue(null);

    const actionData = await action(
      fromPartial({
        request: {
          formData: async () => createMockFormData({ username: '' }),
        },
      })
    );
    vi.mocked(useActionData<typeof action>).mockReturnValue(
      await actionData.json()
    );

    setupRemixStub(<Homepage />);

    const error = screen.getByRole('alert');

    expect(error).toHaveTextContent(/username cannot be empty/i);
  });

  it('renders an error if the username is already taken', async () => {
    vi.mocked(createUser).mockResolvedValue(null);

    const actionData = await action(
      fromPartial({
        request: {
          formData: async () =>
            createMockFormData({ username: 'mock-username' }),
        },
      })
    );
    vi.mocked(useActionData<typeof action>).mockReturnValue(
      await actionData.json()
    );

    setupRemixStub(<Homepage />);

    const error = screen.getByRole('alert');

    expect(error).toHaveTextContent(/username already taken/i);
  });

  it('redirects to the user inbox upon creation', async () => {
    vi.mocked(syncSession).mockResolvedValue('mock-cookie');
    vi.mocked(createUser).mockResolvedValue(fromPartial({ id: 'mock-id' }));

    const response = await action(
      fromPartial({
        request: {
          formData: async () => createMockFormData({ username: 'mock-id' }),
        },
      })
    );

    const expectedHeaders = new Headers();
    expectedHeaders.append('location', '/mock-id');
    expectedHeaders.append('set-cookie', 'mock-cookie');

    expect(response.status).toBe(302);
    expect(response.headers).toEqual(expectedHeaders);
  });

  it.each([
    { name: 'random', text: /generating a random name/i },
    { name: 'create', text: /creating your inbox/i },
  ])(
    'shows a loading icon and disables the $name button on click',
    async ({ name, text }) => {
      vi.mocked(useNavigation).mockReturnValue(
        fromPartial({
          formData: createMockFormData({ _action: name }),
        })
      );

      const { container } = setupRemixStub(<Homepage />);

      expect(container).toMatchSnapshot();
      expect(screen.getByText(text).parentElement).toBeDisabled();
    }
  );

  it('displays any inboxes that are stored in the cookie session', async () => {
    vi.mocked(getInboxes).mockResolvedValue([
      { username: 'mock-inbox-name', date: new Date().toISOString() },
    ]);
    const loaderData = await loader(fromPartial({}));
    vi.mocked(useLoaderData<typeof loader>).mockReturnValue(
      await loaderData.json()
    );

    setupRemixStub(<Homepage />);

    const inbox = screen.getByRole('link');

    expect(inbox).toHaveTextContent('mock-inbox-name@shuttle.email');
  });
});
