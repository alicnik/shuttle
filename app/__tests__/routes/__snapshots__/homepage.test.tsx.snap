// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`homepage > should match the snapshot 1`] = `
<div>
  <div
    class="container mx-auto flex flex-col items-center"
  >
    <h1
      class="mb-2 mt-8 font-display text-6xl font-bold md:mt-16 md:text-8xl"
    >
      Shuttle
    </h1>
    <p
      class="text-center text-lg"
    >
      A temporary email service that minimises the back and forth
    </p>
    <img
      alt="Shuttle Logo"
      class="shuttle-logo cursor-pointer invert dark:invert-0 animate-[pulse_3s_cubic-bezier(0.4,0,0.6,1)_infinite]"
      height="150"
      src="/assets/shuttle.png"
      width="150"
    />
    <form
      action="/"
      class="relative -top-8 mb-4 w-full max-w-sm md:mb-8"
      method="post"
    >
      <p
        class="mb-4 w-full text-center"
      >
        Create an inbox. It will be deleted at midnight.
      </p>
      <div
        class="flex h-9 w-full gap-2 rounded-md border border-input pr-3 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 mb-2"
      >
        <input
          class="flex-1 bg-transparent py-1 pl-3"
          name="username"
          placeholder="Email"
          type="text"
          value=""
        />
        <span
          class="self-center text-muted-foreground"
        >
          @shuttle.email
        </span>
      </div>
       
      <div
        class="flex justify-center gap-4"
      >
        <button
          class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2 w-24"
          name="_action"
          type="submit"
          value="random"
        >
          Random
        </button>
        <button
          class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2 w-24"
          name="_action"
          type="submit"
          value="create"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`homepage > shows a loading icon and disables the 'create' button on click 1`] = `
<div>
  <div
    class="container mx-auto flex flex-col items-center"
  >
    <h1
      class="mb-2 mt-8 font-display text-6xl font-bold md:mt-16 md:text-8xl"
    >
      Shuttle
    </h1>
    <p
      class="text-center text-lg"
    >
      A temporary email service that minimises the back and forth
    </p>
    <img
      alt="Shuttle Logo"
      class="shuttle-logo cursor-pointer invert dark:invert-0 animate-[pulse_3s_cubic-bezier(0.4,0,0.6,1)_infinite]"
      height="150"
      src="/assets/shuttle.png"
      width="150"
    />
    <form
      action="/"
      class="relative -top-8 mb-4 w-full max-w-sm md:mb-8"
      method="post"
    >
      <p
        class="mb-4 w-full text-center"
      >
        Create an inbox. It will be deleted at midnight.
      </p>
      <div
        class="flex h-9 w-full gap-2 rounded-md border border-input pr-3 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 mb-2"
      >
        <input
          class="flex-1 bg-transparent py-1 pl-3"
          name="username"
          placeholder="Email"
          type="text"
          value=""
        />
        <span
          class="self-center text-muted-foreground"
        >
          @shuttle.email
        </span>
      </div>
       
      <div
        class="flex justify-center gap-4"
      >
        <button
          class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2 w-24"
          name="_action"
          type="submit"
          value="random"
        >
          Random
        </button>
        <button
          class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2 w-24"
          disabled=""
          name="_action"
          type="submit"
          value="create"
        >
          <svg
            aria-hidden="true"
            class="mr-2 h-4 w-4 animate-spin"
            fill="none"
            focusable="false"
            height="15"
            viewBox="0 0 15 15"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M1.84998 7.49998C1.84998 4.66458 4.05979 1.84998 7.49998 1.84998C10.2783 1.84998 11.6515 3.9064 12.2367 5H10.5C10.2239 5 10 5.22386 10 5.5C10 5.77614 10.2239 6 10.5 6H13.5C13.7761 6 14 5.77614 14 5.5V2.5C14 2.22386 13.7761 2 13.5 2C13.2239 2 13 2.22386 13 2.5V4.31318C12.2955 3.07126 10.6659 0.849976 7.49998 0.849976C3.43716 0.849976 0.849976 4.18537 0.849976 7.49998C0.849976 10.8146 3.43716 14.15 7.49998 14.15C9.44382 14.15 11.0622 13.3808 12.2145 12.2084C12.8315 11.5806 13.3133 10.839 13.6418 10.0407C13.7469 9.78536 13.6251 9.49315 13.3698 9.38806C13.1144 9.28296 12.8222 9.40478 12.7171 9.66014C12.4363 10.3425 12.0251 10.9745 11.5013 11.5074C10.5295 12.4963 9.16504 13.15 7.49998 13.15C4.05979 13.15 1.84998 10.3354 1.84998 7.49998Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
          <span
            style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
          >
            Creating your inbox
          </span>
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`homepage > shows a loading icon and disables the 'random' button on click 1`] = `
<div>
  <div
    class="container mx-auto flex flex-col items-center"
  >
    <h1
      class="mb-2 mt-8 font-display text-6xl font-bold md:mt-16 md:text-8xl"
    >
      Shuttle
    </h1>
    <p
      class="text-center text-lg"
    >
      A temporary email service that minimises the back and forth
    </p>
    <img
      alt="Shuttle Logo"
      class="shuttle-logo cursor-pointer invert dark:invert-0 animate-[pulse_3s_cubic-bezier(0.4,0,0.6,1)_infinite]"
      height="150"
      src="/assets/shuttle.png"
      width="150"
    />
    <form
      action="/"
      class="relative -top-8 mb-4 w-full max-w-sm md:mb-8"
      method="post"
    >
      <p
        class="mb-4 w-full text-center"
      >
        Create an inbox. It will be deleted at midnight.
      </p>
      <div
        class="flex h-9 w-full gap-2 rounded-md border border-input pr-3 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 mb-2"
      >
        <input
          class="flex-1 bg-transparent py-1 pl-3"
          name="username"
          placeholder="Email"
          type="text"
          value=""
        />
        <span
          class="self-center text-muted-foreground"
        >
          @shuttle.email
        </span>
      </div>
       
      <div
        class="flex justify-center gap-4"
      >
        <button
          class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2 w-24"
          disabled=""
          name="_action"
          type="submit"
          value="random"
        >
          <svg
            aria-hidden="true"
            class="mr-2 h-4 w-4 animate-spin"
            fill="none"
            focusable="false"
            height="15"
            viewBox="0 0 15 15"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M1.84998 7.49998C1.84998 4.66458 4.05979 1.84998 7.49998 1.84998C10.2783 1.84998 11.6515 3.9064 12.2367 5H10.5C10.2239 5 10 5.22386 10 5.5C10 5.77614 10.2239 6 10.5 6H13.5C13.7761 6 14 5.77614 14 5.5V2.5C14 2.22386 13.7761 2 13.5 2C13.2239 2 13 2.22386 13 2.5V4.31318C12.2955 3.07126 10.6659 0.849976 7.49998 0.849976C3.43716 0.849976 0.849976 4.18537 0.849976 7.49998C0.849976 10.8146 3.43716 14.15 7.49998 14.15C9.44382 14.15 11.0622 13.3808 12.2145 12.2084C12.8315 11.5806 13.3133 10.839 13.6418 10.0407C13.7469 9.78536 13.6251 9.49315 13.3698 9.38806C13.1144 9.28296 12.8222 9.40478 12.7171 9.66014C12.4363 10.3425 12.0251 10.9745 11.5013 11.5074C10.5295 12.4963 9.16504 13.15 7.49998 13.15C4.05979 13.15 1.84998 10.3354 1.84998 7.49998Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
          <span
            style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
          >
            Generating a random name
          </span>
        </button>
        <button
          class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2 w-24"
          name="_action"
          type="submit"
          value="create"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;
